### PR TITLE
Cesium3DTileset tests

### PR DESCRIPTION
**Pull request by lilleyse**
For #3177 

* Added tests for `Cesium3DTileset` and `Tileset3DTileContentProvider`
* Changed the replacement refinement approach to work closer with the `selectTiles` function. I'm still not 100% happy with the approach. I was experimenting with other approaches but it wasn't any better.
* Fixed issue where tile with content box is not culled. It was checking the `CullingVolume` mask instead of `Intersect`.
* Changed up the `addToProcessingQueue` and `removeFromProcessingQueue` functions which weren't operating correctly in certain edge cases, like failed requests. This area in general will change again with the request scheduler.
* Implemented `Cesium3DTileset.destroy`. In order to handle cases where the tileset is destroyed while content requests are still in progress, when the request completes I throw an error to reject the ready promise. However I'm open to different ideas.
* For the test tileset files, they all use copies of the same b3dm files which is a bit wasteful. Maybe we could support creating a `Cesium3DTileset` with a url to an explicit `tiles.json` file, right now you can only give it a folder.

### CODE REVIEW

**Comment by pjcozzi** [comment]
> Fixed issue where tile with content box is not culled. It was checking the CullingVolume mask instead of Intersect.

Great, this fixes part of #3241:

>  Does culling with content.box still work? For example, the root tile for Canary Wharf is always rendered when just the tile's box is visible (even it the content box is not).

**Comment by pjcozzi** [comment]
> Maybe we could support creating a Cesium3DTileset with a url to an explicit tiles.json file, right now you can only give it a folder.

Go for it.  Perhaps assume the filename is tiles.json if one isn't provided.

File: Source/Scene/Batched3DModel3DTileContentProvider.js [hunk 1]
```diff
@@ -122,6 +122,9 @@ define([
         this.state = Cesium3DTileContentState.LOADING;
 
         loadArrayBuffer(this._url).then(function(arrayBuffer) {
+            if (that.isDestroyed()) {
+                throw new Error('tileset is destroyed');
```
Comment (pjcozzi) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: We only throw `DeveloperError` or `RuntimeError`, not `Error`; see https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Documentation/Contributors/CodingGuide/README.md#throwing-exceptions

In this case, there was an outstanding request after the tileset was destroyed, right?  That is not an exceptional case.  Can we instead just silently destroy the tile?

Comment (lilleyse) [inline Source/Scene/Batched3DModel3DTileContentProvider.js]: Yeah it can be silent. I did it this way so that the request scheduler and stats would get updated correctly, but since that all is changing anyways I think it's fine not to throw the errors.

File: Source/Scene/Cesium3DTileset.js [hunk 2]
```diff
@@ -542,18 +548,16 @@ define([
...
-            --tiles3D._statistics.numberProcessing;
-            addLoadProgressEvent(tiles3D);
-        };
-    }
+            if (index >= 0) {
+                // Remove from processing queue
+                tiles3D._processingQueue.splice(index, 1);
+                --tiles3D._statistics.numberProcessing;
+            } else {
+                // Not in processing queue
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Can we add a comment on why it wasn't in the queue?

File: Specs/Data/Cesium3DTiles/Tilesets/TilesetReplacement2/tiles.json [hunk 3]
```diff
@@ -0,0 +1,80 @@
+{
+    "properties": {
+        "id": {
+            "minimum": 0,
+            "maximum": 99
+        }
+    },
+    "geometricError": 240,
+    "refine": "add",
```
Comment (pjcozzi) [inline Specs/Data/Cesium3DTiles/Tilesets/TilesetReplacement2/tiles.json]: There is not top-level `refine` property.  This should be part of the tile.

**Comment by pjcozzi** [comment]
Do we have any test cases for both additive and replacement refinement in the same tileset?

**Comment by lilleyse** [comment]
`TileReplacement2` has a mix, but I do think I should add a more general case test.

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+            expect(stats.visited).toEqual(0); // selectTiles returns early, so no tiles are visited
+            expect(stats.numberOfCommands).toEqual(1); // root tile is still in selectedTiles list
+        });
+    });
+
+    it('debugShowStatistics', function() {
+        return Cesium3DTilesTester.loadTileset(scene, tilesetUrl).then(function(tileset) {
+            tileset.debugShowStatistics = true;
+            scene.renderForSpecs();
+            // TODO : not sure how to test this
```
Comment (pjcozzi) [inline Specs/Scene/Cesium3DTilesetSpec.js]: Use a spy for console log.  Don't too tightly couple this test to the implementation though, just a coarse-grained test is plenty.

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+        });
+    });
+
+    it('does not request tiles when out of core', function() {
```
Comment (pjcozzi) [inline Specs/Scene/Cesium3DTilesetSpec.js]: Here and below: "when picking", not "when out of core."  It is _not_ out of core.

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+        });
+    });
+
+    it('destroys before loadTilesJson finishes', function() {
+        var tileset = scene.primitives.add(new Cesium3DTileset({
+            url : tilesetUrl
+        }));
+        scene.primitives.remove(tileset);
+        // TODO : hard to test this one (check that promise never resolves or rejects?)
+        // Should destroy function reject ready promise?
```
Comment (pjcozzi) [inline Specs/Scene/Cesium3DTilesetSpec.js]: The behavior is up to you, but it should be consistent with `Model` (changes to `Model` are OK).

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+            var stats = tileset._statistics;
+            expect(stats.numberOfPendingRequests).toEqual(1);
+            scene.primitives.remove(tileset);
+
+            return root.readyPromise.then(function(root) {
+                fail('should not resolve');
+            }).otherwise(function(error) {
+                // Expect the root to not have added any children from the external tiles.json
+                expect(root.children.length).toEqual(0);
+                // TODO : check that request scheduler is empty
```
Comment (pjcozzi) [inline Specs/Scene/Cesium3DTilesetSpec.js]: We can provide a private accessor for now to test this.

**Comment by pjcozzi** [comment]
The `debug*` options, such as `debugShowBox` need tests.  As usual, coarse-grained checks are fine.

**Comment by pjcozzi** [comment]
Can we add an explicit test for:

> // Use parent's geometric error with child's box to see if we already meet the SSE

https://github.com/AnalyticalGraphicsInc/cesium/blob/tileset-tests/Source/Scene/Cesium3DTileset.js#L461

This is an important optimization for additive refinement.

Basically, from one view we should not need to request the farther away children, then move the view closer and the children are requested.

**Comment by pjcozzi** [comment]
Can we add a test to make sure tiles are traversed (requested/rendered) in the a front-to-back order?

I think there is an issue here.  From #3241:

>  Is the load order correct? For example, the buildings on the right should load sooner: http://cesiumjs.org/WashingtonDC/

We can also punt on this until the request scheduler if that makes more sense.

**Comment by pjcozzi** [comment]
Let's test the case when the content url exists but it has an unknown extension, e.g.,
```
            "content": {
              "url": "2/0/0.unknown"
            },
```

**Comment by pjcozzi** [comment]
Do we need a test case for a traditional quadtree where all the interior tiles are empty and only the leafs have content?  Or is `TilesetEmptyRoot` enough?

**Comment by pjcozzi** [comment]
Would some of these tests be better if we mocked out (or just spied on) `Cesium3DTile` and friends?

It would make the tests more precise and depend on fewer moving parts.

**Comment by pjcozzi** [comment]
> Would some of these tests be better if we mocked out (or just spied on) Cesium3DTile and friends?

> It would make the tests more precise and depend on fewer moving parts.

Also, this is up to you if you think it is cleaner.

**Comment by pjcozzi** [comment]
Did coverage bump to 94%? :)

**Comment by lilleyse** [comment]
Not yet...

**Comment by pjcozzi** [comment]
That's all my comments.  This looks really good.

**Comment by lilleyse** [comment]
> Let's test the case when the content url exists but it has an unknown extension, e.g.,

This will be easier to test in `Cesium3DTileSpec`

File: Specs/Scene/Cesium3DTilesetSpec.js (hunk 7 above)
Comment (lilleyse) [inline Specs/Scene/Cesium3DTilesetSpec.js]: I may just leave them there for now and wait until the request scheduler is ready.

**Comment by lilleyse** [comment]
This is ready for another look.

**Comment by pjcozzi** [comment]
Coverage for `Cesium3DTile` is not great, but it sounds like you plan to handle that in a separate PR.
